---
title: "Microbiome analysis using phyloseq"
---
This R Markdown contains a starting collection of basic microbiome analysis using pyhloseq (https://github.com/joey711/phyloseq). A tutorial can be accessed here: https://joey711.github.io/phyloseq/import-data.html

###1. Set up - install packages and load functions
If you have not yet installed thi packages => first install (usig https://github.com/janinazuend/Analysis-16S-rRNA/blob/main/setup.Rmd)
```{r setup, include=FALSE}
mylib = 'C:\\Users\\zuendj\\Documents\\R4Libs'
.libPaths(mylib)

# devtools::install_github("tidyverse/tidyverse") 

# library(phyloseq)
# library(tidyverse)
# library(openxlsx)
# library(microViz)
# library(rstatix)
# library(ggpubr)
# library(plyr)
# library(ggpmisc)
# library(cowplot)
# library(vegan)
```

```{r}
#download the functions and store them locally on your computer=> from fconstancias: https://zenodo.org/records/6473394 
#load them into R => replace directory with the directory on your computer
source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_taxa_tests.R") 
source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_normalisation.R") 
source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_alpha.R") 
source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_beta.R") 
source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_heatmap.R")
source("C:/Users/zuendj/Desktop/03_data R/divcom_functions/R/phyloseq_taxa_tests.R")

#function for getting metadata file
sample.data.frame <- function(ps) {
  return(as(phyloseq::sample_data(ps), "data.frame"))}

#function to name ggplot facets
facet_labeller <- function(variable,value){
  return(names[value])
}
```

### 1. Load/create a phyloseq object 
For different methods to create phyloseq see: https://joey711.github.io/phyloseq/import-data.html

```{r}
ps<-readRDS("C:/Users/zuendj/Desktop/03_data R/fibreGum/phyloseq_FibreGum_Run1.rds")
```

### 2. Check the Library size & Rarefaction (Normalization => for Diversity analysis !!!)
```{r}
# Extract the OTU (or ASV) table as a matrix
otu_mat <- as(otu_table(ps), "matrix")

# If taxa are rows, transpose to have samples in rows
if(taxa_are_rows(ps)) {
  otu_mat <- t(otu_mat)
}

# Calculate library size per sample (sum of counts per sample)
library_sizes <- setNames(as.data.frame(rowSums(otu_mat)), c("Libsize"))%>%
  arrange(Libsize) ##order
```
```{r}
#plot the read depth
vegan::rarecurve(otu_mat, step=10000, cex=0.5)
```
#### Rarefying the reads
```{r}
ps %>%
  rarefy_even_depth(rngseed = 123,
                    sample.size = 1095326
                    ) -> ps_rare
```
###  2. Alpha diversity
```{r}
#alpha diversity
ps_rare%>%
  phyloseq_alphas(phylo = F) -> alphas
```

```{r}
alphas %>%
  gather(index, alpha_value, "observed":"diversity_shannon") %>%
  mutate(
    Type = factor(Type, levels = c("Fecal", "Saliva", "Buccal")),
    Arm = factor(Arm, levels = c("Intervention", "Ctrl")),
    index = factor(index, levels = c("observed", "diversity_shannon"))
  ) %>%
  
  ggplot(aes(x = Arm, y = alpha_value, fill = Time)) +
  geom_boxplot(position = position_dodge(width = 0.75), outlier.shape = NA) +
  geom_point(aes(color = Time),
             position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.75),
             alpha = 0.75, size = 3) +
  ylab("Diversity") + 
  xlab(NULL) +
  facet_grid(rows = vars(index), cols = vars(Type), scales = "free", space = "free_x") +
  theme(text = element_text(size = 20),
    panel.background = element_rect(fill = "white"),
    panel.border = element_rect(fill = NA)
  )+
scale_fill_manual(values=c("white", "white"))+
scale_color_manual(values=c("seagreen3", "lightpink"))

ggsave(filename = "alpha_run1.jpeg",
       path ="C:/Users/zuendj/Desktop/03_data R/fibreGum/output",
       width = 20,
       height = 17,
       units = c("cm"))
```

### 3. Compositional bar plots using Microviz
```{r}
ps_rare -> ps_rare_fix
tax_mat <- as.matrix(tax_table(ps_rare_fix))
tax_mat <- apply(tax_mat, 2, as.character)  # Sicherstellen, dass alles Zeichen ist
rownames(tax_mat) <- taxa_names(ps_rare_fix)

for (i in 1:nrow(tax_mat)) {
  for (j in 2:ncol(tax_mat)) {  # von 2 bis 7 (bzw. Anzahl der Spalten)
    if (is.na(tax_mat[i, j]) || tax_mat[i, j] == "") {  # NA oder leer ersetzen
      higher_ranks <- tax_mat[i, 1:(j - 1)]
      fallback <- tail(na.omit(higher_ranks[higher_ranks != ""]), n = 1)
      if (length(fallback) > 0) {
        tax_mat[i, j] <- fallback
      }
    }
  }
}



# Assign the modified matrix back to the phyloseq object
tax_table(ps_rare_fix) <- tax_table(tax_mat)

ps_rare_fix%>%
  tax_fix()->ps_fix
```

```{r}
meta<-sample.data.frame(ps_fix)



sample_data(ps_fix)<-meta%>%
  mutate(Time = factor(Time, levels = c("Base", "End")))
```

```{r}
ps_fix%>%
  ps_arrange(Time)%>%
  microViz::comp_barplot(
    tax_level = "Family",
    label = "Time",
    n_taxa = 10,
    bar_width = 0.9,
    sample_order = "default",
    tax_transform_for_plot = "compositional"
    # palette= myPal
    ) +
    ylab("Proportion") + xlab("")-> p_hist

p_hist + facet_grid(rows=vars(Type, UPN), scales = "free", space = "free_x", drop = TRUE)+
  theme(text= element_text(size=15),
        axis.text.y=element_text(angle=0),
        axis.text.x = element_text(vjust=0.2, hjust=1, margin=margin(-8,0,0,0)),
        axis.ticks.x = element_blank(),
        legend.text = element_text(vjust=1, hjust=0),
        panel.spacing = unit(0.05, "cm"),
        strip.text.y = element_text(size=15, angle=0)) + 
  scale_y_continuous(breaks=c(0.25, 0.5, 0.75))+
  coord_flip()

ggsave(filename = "Barplots_family_Fibregum_Run_1.png",
       path = "C:/Users/zuendj/Desktop/03_data R/fibreGum/output",
       width = 20,
       height = 35,
       units = c("cm"))
```

